File: src/main/java/so/aporia/utils/user/render/core/AporiaRenderer.java
package so.aporia.utils.user.render.core;

import com.mojang.blaze3d.ProjectionType;
import com.mojang.blaze3d.buffers.GpuBuffer;
import com.mojang.blaze3d.pipeline.BlendFunction;
import com.mojang.blaze3d.pipeline.RenderPipeline;
import com.mojang.blaze3d.pipeline.TextureTarget;
import com.mojang.blaze3d.platform.DepthTestFunction;
import com.mojang.blaze3d.shaders.UniformType;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.textures.FilterMode;
import com.mojang.blaze3d.vertex.DefaultVertexFormat;
import com.mojang.blaze3d.vertex.Tesselator;
import com.mojang.blaze3d.vertex.VertexFormat;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.CachedOrthoProjectionMatrixBuffer;
import net.minecraft.resources.Identifier;
import so.aporia.utils.user.render.font.Fonts;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.OptionalInt;

public class AporiaRenderer {

    public static final AporiaRenderer INSTANCE = new AporiaRenderer();

    /** Mode constants — must match values in {@code aporia.fsh}. */
    public static final int MODE_FILL         = 0;
    public static final int MODE_CIRCLE       = 1;
    public static final int MODE_ROUNDED_RECT = 2;

    private RenderPipeline pipeline;
    private CachedOrthoProjectionMatrixBuffer orthoProjection;

    /* Blur pipeline fields */
    private RenderPipeline blurPipeline;
    private RenderPipeline blitPipeline;
    private TextureTarget blurTarget;
    private GpuBuffer blurUniformBuffer;
    
    /* Blur rectangle settings */
    private float blurRectW = 300f;
    private float blurRectH = 200f;
    private float blurRectRadius = 15f;
    private float blurRectSmoothness = 2f;
    private float blurRadius = 15f;

    public void init() {
        pipeline = RenderPipeline.builder()
            .withLocation(Identifier.fromNamespaceAndPath("aporia", "pipeline/aporia"))
            .withVertexShader(Identifier.fromNamespaceAndPath("aporia", "core/aporia"))
            .withFragmentShader(Identifier.fromNamespaceAndPath("aporia", "core/aporia"))
            .withUniform("Projection", com.mojang.blaze3d.shaders.UniformType.UNIFORM_BUFFER)
            .withUniform("ShapeData",  com.mojang.blaze3d.shaders.UniformType.UNIFORM_BUFFER)
            .withVertexFormat(DefaultVertexFormat.POSITION_TEX_COLOR, VertexFormat.Mode.TRIANGLES)
            .withBlend(BlendFunction.TRANSLUCENT)
            .withDepthTestFunction(DepthTestFunction.NO_DEPTH_TEST)
            .withDepthWrite(false)
            .withCull(false)
            .build();

        orthoProjection = new CachedOrthoProjectionMatrixBuffer("aporia", -1000f, 1000f, true);

        /* Init blur pipeline */
        blurPipeline = RenderPipeline.builder()
            .withLocation(Identifier.fromNamespaceAndPath("aporia", "pipeline/blur"))
            .withVertexShader(Identifier.fromNamespaceAndPath("aporia", "core/blur"))
            .withFragmentShader(Identifier.fromNamespaceAndPath("aporia", "core/blur"))
            .withSampler("InputTexture")
            .withVertexFormat(DefaultVertexFormat.POSITION_COLOR, VertexFormat.Mode.TRIANGLES)
            .withBlend(BlendFunction.TRANSLUCENT)
            .withDepthTestFunction(DepthTestFunction.NO_DEPTH_TEST)
            .withDepthWrite(false)
            .withCull(false)
            .build();

        /* Init blit pipeline - uses simple blit shader */
        blitPipeline = RenderPipeline.builder()
            .withLocation(Identifier.fromNamespaceAndPath("aporia", "pipeline/blur_blit"))
            .withVertexShader(Identifier.fromNamespaceAndPath("aporia", "core/blit"))
            .withFragmentShader(Identifier.fromNamespaceAndPath("aporia", "core/blit"))
            .withSampler("InputTexture")
            .withVertexFormat(DefaultVertexFormat.POSITION_TEX, VertexFormat.Mode.QUADS)
            .withBlend(BlendFunction.TRANSLUCENT)
            .withDepthTestFunction(DepthTestFunction.NO_DEPTH_TEST)
            .withDepthWrite(false)
            .withCull(false)
            .build();
    }

    /** Public drawing API. */

    /** Line from (x1,y1) to (x2,y2) with given pixel thickness */
    public void drawLine(float x1, float y1, float x2, float y2, float thickness, int color) {
        float dx = x2 - x1, dy = y2 - y1;
        float len = (float) Math.sqrt(dx * dx + dy * dy);
        if (len == 0) return;
        float nx = -dy / len * thickness * 0.5f;
        float ny =  dx / len * thickness * 0.5f;

        /* Two triangles forming a quad along the line. */
        float[] vx = { x1+nx, x1-nx, x2-nx, x2+nx };
        float[] vy = { y1+ny, y1-ny, y2-ny, y2+ny };

        /* Bounds = line bounding box for SDF (unused in mode 0, but populated for consistency). */
        float bx = Math.min(x1, x2) - thickness;
        float by = Math.min(y1, y2) - thickness;
        float bw = Math.abs(dx) + thickness * 2;
        float bh = Math.abs(dy) + thickness * 2;

        draw(new float[][]{
            {vx[0], vy[0]}, {vx[1], vy[1]}, {vx[2], vy[2]},
            {vx[0], vy[0]}, {vx[2], vy[2]}, {vx[3], vy[3]}
        }, color, MODE_FILL, bx, by, bw, bh, 0f);
    }

    /** Filled circle */
    public void drawCircle(float cx, float cy, float radius, int color) {
        float x = cx - radius, y = cy - radius, d = radius * 2;
        drawShape(x, y, d, d, color, MODE_CIRCLE, x, y, d, d, radius, 0, 0f, 0f);
    }

    /** Filled triangle with 3 explicit vertices */
    public void drawTriangle(float x1, float y1, float x2, float y2, float x3, float y3, int color) {
        float bx = Math.min(x1, Math.min(x2, x3));
        float by = Math.min(y1, Math.min(y2, y3));
        float bw = Math.max(x1, Math.max(x2, x3)) - bx;
        float bh = Math.max(y1, Math.max(y2, y3)) - by;
        draw(new float[][]{{x1,y1},{x2,y2},{x3,y3}}, color, MODE_FILL, bx, by, bw, bh, 0f);
    }

    /** Filled rectangle, optionally rounded */
    public void drawRect(float x, float y, float w, float h, float radius, int color) {
        int mode = radius > 0 ? MODE_ROUNDED_RECT : MODE_FILL;
        drawShape(x, y, w, h, color, mode, x, y, w, h, radius, 0, 0f, 0f);
    }

    /**
     * Rounded rectangle stroke.
     * {@code borderMode} — 1=full outline, 2=corners-only.
     * {@code thickness}  — stroke width in pixels.
     * {@code fadeCorner} — 0..1, how much straight edges fade (only for mode 2).
     */
    public void drawStroke(float x, float y, float w, float h, float radius,
                           float thickness, int borderMode, float fadeCorner, int color) {
        drawShape(x, y, w, h, color, MODE_ROUNDED_RECT, x, y, w, h, radius,
                  borderMode, thickness, fadeCorner);
    }

    /** Internal helpers. */

    private void drawShape(float x, float y, float w, float h, int color,
                           int mode, float bx, float by, float bw, float bh,
                           float radius, int borderMode, float thickness, float fadeCorner) {
        draw(new float[][]{
            {x,   y+h}, {x+w, y+h}, {x+w, y},
            {x,   y+h}, {x+w, y  }, {x,   y}
        }, color, mode, bx, by, bw, bh, radius, borderMode, thickness, fadeCorner);
    }

    private void draw(float[][] verts, int color, int mode,
                      float bx, float by, float bw, float bh, float radius) {
        draw(verts, color, mode, bx, by, bw, bh, radius, 0, 0f, 0f);
    }

    private void draw(float[][] verts, int color, int mode,
                      float bx, float by, float bw, float bh, float radius,
                      int borderMode, float thickness, float fadeCorner) {
        Minecraft mc = Minecraft.getInstance();
        var window    = mc.getWindow();
        var colorView = mc.getMainRenderTarget().getColorTextureView();
        if (colorView == null || pipeline == null) return;

        float a = ((color >> 24) & 0xFF) / 255f;
        float r = ((color >> 16) & 0xFF) / 255f;
        float g = ((color >>  8) & 0xFF) / 255f;
        float b = ((color      ) & 0xFF) / 255f;

        var projSlice = orthoProjection.getBuffer(window.getGuiScaledWidth(), window.getGuiScaledHeight());
        RenderSystem.setProjectionMatrix(projSlice, ProjectionType.ORTHOGRAPHIC);

        var tess = Tesselator.getInstance();
        var buf  = tess.begin(VertexFormat.Mode.TRIANGLES, DefaultVertexFormat.POSITION_TEX_COLOR);
        for (float[] v : verts) {
            buf.addVertex(v[0], v[1], 0f).setUv(0f, 0f).setColor(r, g, b, a);
        }
        var mesh = buf.buildOrThrow();

        var device  = RenderSystem.getDevice();
        var encoder = device.createCommandEncoder();

        var vertexGpu = device.createBuffer(
            () -> "aporia:vbo",
            GpuBuffer.USAGE_VERTEX | GpuBuffer.USAGE_COPY_DST,
            mesh.vertexBuffer()
        );

        /* ShapeData UBO layout: bounds(4) + params(4) + params2(4) = 48 bytes */
        var shapeBuf = device.createBuffer(() -> "aporia:shape", GpuBuffer.USAGE_UNIFORM | GpuBuffer.USAGE_COPY_DST, 48L);
        var bb = ByteBuffer.allocateDirect(48).order(ByteOrder.nativeOrder());
        bb.putFloat(bx); bb.putFloat(by); bb.putFloat(bw); bb.putFloat(bh);       /* bounds  */
        bb.putFloat(radius); bb.putFloat(1.0f); bb.putFloat(mode); bb.putFloat((float) borderMode); /* params  */
        bb.putFloat(thickness); bb.putFloat(fadeCorner); bb.putFloat(0f); bb.putFloat(0f);           /* params2 */
        bb.flip();
        encoder.writeToBuffer(shapeBuf.slice(), bb);

        var indexBuf = RenderSystem.getSequentialBuffer(VertexFormat.Mode.TRIANGLES);

        try (var pass = encoder.createRenderPass(() -> "aporia:draw", colorView, OptionalInt.empty())) {
            pass.setPipeline(pipeline);
            RenderSystem.bindDefaultUniforms(pass);
            pass.setUniform("ShapeData", shapeBuf.slice());
            pass.setVertexBuffer(0, vertexGpu);
            pass.setIndexBuffer(indexBuf.getBuffer(verts.length), indexBuf.type());
            pass.drawIndexed(0, 0, verts.length / 3 * 3, 0);
        }

        mesh.close();
        vertexGpu.close();
        shapeBuf.close();
    }

    /** Test render entry point — shapes are commented out until needed. */

    public void onRenderHud(Minecraft mc) {
    }

    /** Draws text via the shared {@link so.aporia.utils.user.render.font.FontRenderer}. */
    public void drawText(String font, String text, float x, float y, float size, int color) {
        so.aporia.Aporia.FONTS.drawText(font, text, x, y, size, color);
    }

    /** Draws text with an MSDF outline. */
    public void drawTextWithOutline(String font, String text, float x, float y, float size,
                                    int color, float outlineWidth, int outlineColor) {
        so.aporia.Aporia.FONTS.drawTextWithOutline(font, text, x, y, size, color, outlineWidth, outlineColor);
    }

    /** Returns pixel width of text at given size. */
    public float getTextWidth(String font, String text, float size) {
        return so.aporia.Aporia.FONTS.getTextWidth(font, text, size);
    }

    /**
     * Draws a single glyph from the big icon font by index.
     * Index maps to PUA codepoint {@code 0xE000 + index}.
     */
    public void drawGlyph(int index, float x, float y, float size, int color) {
        so.aporia.Aporia.FONTS.drawGlyph(Fonts.FONT, index, x, y, size, color);
    }

    public void onRenderWorld(Minecraft mc) {}

    /**
     * Applies blur effect to a rectangular region.
     * Called from Blur module each frame when enabled.
     */
    public void applyBlur(Minecraft mc, float blurRadius) {
        var mainTarget = mc.getMainRenderTarget();
        var colorView = mainTarget.getColorTextureView();
        if (colorView == null || blurPipeline == null) return;

        var device = RenderSystem.getDevice();
        var encoder = device.createCommandEncoder();

        /* Draw a quad with blur effect in center of screen */
        var tess = Tesselator.getInstance();
        var buf = tess.begin(VertexFormat.Mode.TRIANGLES, DefaultVertexFormat.POSITION_COLOR);
        
        /* Center quad - NDC coordinates */
        float size = 0.4f;
        buf.addVertex(-size, size, 0f).setColor(1f, 1f, 1f, 1f);
        buf.addVertex(-size, -size, 0f).setColor(1f, 1f, 1f, 1f);
        buf.addVertex(size, -size, 0f).setColor(1f, 1f, 1f, 1f);
        buf.addVertex(-size, size, 0f).setColor(1f, 1f, 1f, 1f);
        buf.addVertex(size, -size, 0f).setColor(1f, 1f, 1f, 1f);
        buf.addVertex(size, size, 0f).setColor(1f, 1f, 1f, 1f);

        var mesh = buf.buildOrThrow();
        var vertexGpu = device.createBuffer(
            () -> "aporia:blur_vbo",
            GpuBuffer.USAGE_VERTEX | GpuBuffer.USAGE_COPY_DST,
            mesh.vertexBuffer()
        );

        try (var pass = encoder.createRenderPass(() -> "aporia:blur_pass", mainTarget.getColorTextureView(), OptionalInt.empty())) {
            pass.setPipeline(blurPipeline);
            pass.bindTexture("InputTexture", colorView, RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
            pass.setVertexBuffer(0, vertexGpu);
            pass.draw(0, 6);
        }

        mesh.close();
        vertexGpu.close();
    }

    /**
     * Renders blur effect for the Blur module.
     * Call this from RenderHudEvent when Blur is enabled.
     */
    public void renderBlur(Minecraft mc) {
        applyBlur(mc, 5f);
    }

    /**
     * Cleans up blur resources when module is disabled.
     */
    public void cleanupBlur() {
        if (blurTarget != null) {
            blurTarget.destroyBuffers();
            blurTarget = null;
        }
        if (blurUniformBuffer != null) {
            blurUniformBuffer.close();
            blurUniformBuffer = null;
        }
    }
}

File: src/so/aporia/utils/user/render/core/AporiaRenderer.kt
package so.aporia.utils.user.render.core

import com.mojang.blaze3d.buffers.GpuBuffer
import com.mojang.blaze3d.pipeline.BlendFunction
import com.mojang.blaze3d.pipeline.RenderPipeline
import com.mojang.blaze3d.platform.DepthTestFunction
import com.mojang.blaze3d.shaders.UniformType
import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.textures.GpuTextureView
import com.mojang.blaze3d.vertex.BufferBuilder
import com.mojang.blaze3d.vertex.DefaultVertexFormat
import com.mojang.blaze3d.vertex.VertexFormat
import net.minecraft.client.Minecraft
import net.minecraft.resources.Identifier
import so.aporia.Aporia
import so.aporia.utils.user.logger.Logger
import org.joml.Matrix4f
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Path
import java.util.OptionalDouble
import java.util.OptionalInt

object AporiaRenderer {

    // ── Sub-renderers ──
    val shapes = ShapesRenderer()
    val pixels = PixelsRenderer()

    // ── 3D projection matrices (set by GameRenderer) ──
    @JvmField var worldProjMatrix = Matrix4f()
    @JvmField var worldViewMatrix = Matrix4f()

    // ── Entity glow ──
    private var entityGlowPipeline: RenderPipeline? = null
    private var entityGlowVBO: GpuBuffer? = null
    private var entityGlowMVP: GpuBuffer? = null
    private var entityGlowParams: GpuBuffer? = null
    private var entityGlowIBO: GpuBuffer? = null

    // Pre-allocated scratch buffers
    private var scratchMVP = ByteBuffer.allocateDirect(192).order(ByteOrder.nativeOrder())
    private var scratchParams = ByteBuffer.allocateDirect(32).order(ByteOrder.nativeOrder())
    private var scratchVertex = ByteBuffer.allocateDirect(262144).order(ByteOrder.nativeOrder())
    private var scratchIndexData = ByteBuffer.allocateDirect(12000).order(ByteOrder.nativeOrder())
    private val scratchProj = Matrix4f()
    private val scratchIdentity = Matrix4f()
    private val scratchCombined = Matrix4f()

    // ── Post-process ──
    var saturation = 0.5f

    // ── Depth sorting ──
    fun depth(z: Float) { shapes.depth(z) }
    fun flush() { shapes.flush() }

    // ── Init ──

    fun init() {
        val device = RenderSystem.getDevice()
        shapes.init()
        pixels.init()
        BlurRenderer.init()

        entityGlowPipeline = RenderPipeline.builder()
            .withLocation(Identifier.fromNamespaceAndPath("aporia", "pipeline/entity_glow"))
            .withVertexShader(Identifier.fromNamespaceAndPath("aporia", "core/entity_glow"))
            .withFragmentShader(Identifier.fromNamespaceAndPath("aporia", "core/entity_glow"))
            .withUniform("ModelViewProj", UniformType.UNIFORM_BUFFER)
            .withUniform("DrawParams", UniformType.UNIFORM_BUFFER)
            .withVertexFormat(DefaultVertexFormat.NEW_ENTITY, VertexFormat.Mode.TRIANGLES)
            .withBlend(BlendFunction.TRANSLUCENT)
            .withDepthTestFunction(DepthTestFunction.NO_DEPTH_TEST)
            .withDepthWrite(false).withCull(false).build()

        entityGlowVBO = device.createBuffer({ -> "aporia:entity_glow_vbo" },
            GpuBuffer.USAGE_VERTEX or GpuBuffer.USAGE_COPY_DST, 262144L)
        entityGlowMVP = device.createBuffer({ -> "aporia:entity_glow_mvp" },
            GpuBuffer.USAGE_UNIFORM or GpuBuffer.USAGE_COPY_DST, 192L)
        entityGlowParams = device.createBuffer({ -> "aporia:entity_glow_params" },
            GpuBuffer.USAGE_UNIFORM or GpuBuffer.USAGE_COPY_DST, 32L)
        val maxQuads = 2000
        val maxIndices = maxQuads * 6
        entityGlowIBO = device.createBuffer({ -> "aporia:entity_glow_ibo" },
            GpuBuffer.USAGE_INDEX or GpuBuffer.USAGE_COPY_DST, (maxIndices * 2L))
    }

    // ═══════════════════════════
    //  Shapes — delegated
    // ═══════════════════════════

    fun drawRectBlurred(x: Float, y: Float, w: Float, h: Float, radius: Float, color: Int) =
        shapes.drawRectBlurred(x, y, w, h, radius, color, 4f, 15)
    fun drawRectBlurred(x: Float, y: Float, w: Float, h: Float, radius: Float, color: Int, blurStrength: Float) =
        shapes.drawRectBlurred(x, y, w, h, radius, color, blurStrength, 15)
    fun drawRectBlurred(x: Float, y: Float, w: Float, h: Float, radius: Float, color: Int, blurStrength: Float, cornerMask: Int) =
        shapes.drawRectBlurred(x, y, w, h, radius, color, blurStrength, cornerMask)
    @JvmOverloads fun drawRect(x: Float, y: Float, w: Float, h: Float, radius: Float = 0f, color: Int = -1, cornerMask: Int = 15) =
        shapes.drawRect(x, y, w, h, radius, color, cornerMask)
    fun drawLine(x1: Float, y1: Float, x2: Float, y2: Float, thickness: Float, color: Int) =
        shapes.drawLine(x1, y1, x2, y2, thickness, color)
    fun drawFadeHLine(cx: Float, cy: Float, halfLen: Float, thickness: Float, progress: Float, color: Int) =
        shapes.drawFadeHLine(cx, cy, halfLen, thickness, progress, color)
    fun drawCircle(cx: Float, cy: Float, radius: Float, color: Int) =
        shapes.drawCircle(cx, cy, radius, color)
    fun drawTriangle(x1: Float, y1: Float, x2: Float, y2: Float, x3: Float, y3: Float, color: Int) =
        shapes.drawTriangle(x1, y1, x2, y2, x3, y3, color)
    fun drawStroke(x: Float, y: Float, w: Float, h: Float, radius: Float, thickness: Float, borderMode: Int, fadeCorner: Float, color: Int) =
        shapes.drawStroke(x, y, w, h, radius, thickness, borderMode, fadeCorner, color)
    fun drawMainMenuBackground(time: Float, width: Int, height: Int) =
        shapes.drawMainMenuBackground(time, width, height)
    fun drawLogo(x: Float, y: Float, w: Float, h: Float, tex: Identifier?, time: Long) {
        if (tex != null) shapes.drawLogo(x, y, w, h, tex, time)
    }
    fun drawLogo(x: Float, y: Float, w: Float, h: Float, view: GpuTextureView?, time: Long) =
        shapes.drawLogo(x, y, w, h, view, time)

    // ═══════════════════════════
    //  Pixels — delegated
    // ═══════════════════════════

    fun drawImage(x: Float, y: Float, w: Float, h: Float, id: Identifier?, radius: Float = 0f) =
        pixels.drawImage(x, y, w, h, id, radius)
    fun drawTexture(x: Float, y: Float, w: Float, h: Float, view: GpuTextureView?) =
        pixels.drawTexture(x, y, w, h, view)
    fun drawImageCropped(x: Float, y: Float, w: Float, h: Float, id: Identifier?, radius: Float, u0: Float, v0: Float, u1: Float, v1: Float) =
        pixels.drawImageCropped(x, y, w, h, id, radius, u0, v0, u1, v1)
    fun blitCropped(x: Float, y: Float, w: Float, h: Float, id: Identifier?, u0: Float, v0: Float, u1: Float, v1: Float) =
        pixels.blitCropped(x, y, w, h, id, u0, v0, u1, v1)
    fun applySaturation(mc: Minecraft, saturation: Float) = pixels.applySaturation(saturation)
    fun loadImage(path: Path): Identifier? = pixels.loadImage(path)
    fun loadImage(stream: InputStream): Identifier? = pixels.loadImage(stream)
    @JvmStatic fun cleanupImages() = pixels.cleanupImages()

    // ═══════════════════════════
    //  Fonts
    // ═══════════════════════════

    fun drawText(font: String, text: String, x: Float, y: Float, size: Float, color: Int) =
        Aporia.FONTS.drawText(font, text, x, y, size, color)
    fun getTextWidth(font: String, text: String, size: Float): Float =
        Aporia.FONTS.getTextWidth(font, text, size)

    // ═══════════════════════════
    //  Entity glow (3D)
    // ═══════════════════════════

    fun drawPlayerGlow(
        poseStack: com.mojang.blaze3d.vertex.PoseStack,
        model: net.minecraft.client.model.EntityModel<*>,
        packedLight: Int, packedOverlay: Int,
        colorR: Float, colorG: Float, colorB: Float, colorA: Float,
        glowIntensity: Float, rimPower: Float, fresnelPower: Float, pulseSpeed: Float
    ) {
        if (entityGlowPipeline == null) { Logger.warn("entityGlowPipeline is null"); return }
        val bb = com.mojang.blaze3d.vertex.ByteBufferBuilder(262144)
        val buf = BufferBuilder(bb, VertexFormat.Mode.QUADS, DefaultVertexFormat.NEW_ENTITY)
        model.renderToBuffer(poseStack, buf, packedLight, packedOverlay, -1)
        val mesh = buf.build() ?: run { bb.close(); return }
        val totalBytes = mesh.vertexBuffer().remaining()
        val rawData = ByteArray(totalBytes)
        mesh.vertexBuffer().get(rawData)
        val vertexCount = mesh.drawState().vertexCount()
        mesh.close(); bb.close()
        drawPlayerGlowSubmit(rawData, vertexCount, colorR, colorG, colorB, colorA,
            glowIntensity, rimPower, fresnelPower, pulseSpeed)
    }

    fun drawPlayerGlowSubmit(
        vertexData: ByteArray, vertexCount: Int,
        colorR: Float, colorG: Float, colorB: Float, colorA: Float,
        glowIntensity: Float, rimPower: Float, fresnelPower: Float, pulseSpeed: Float
    ) {
        val pipeline = entityGlowPipeline ?: run { Logger.warn("drawPlayerGlowSubmit: pipeline null"); return }
        val mc2 = Minecraft.getInstance()
        val colorView = mc2.mainRenderTarget.colorTextureView ?: return
        val depthView = mc2.mainRenderTarget.depthTextureView
        if (vertexCount == 0) { Logger.warn("drawPlayerGlowSubmit: vertexCount=0, skipping"); return }

        if (scratchVertex.capacity() < vertexData.size)
            scratchVertex = ByteBuffer.allocateDirect(vertexData.size).order(ByteOrder.nativeOrder())
        scratchVertex.clear(); scratchVertex.put(vertexData); scratchVertex.flip()

        val camPos = mc2.gameRenderer.mainCamera.position()
        val mv = Matrix4f(worldViewMatrix)
        mv.translate(-camPos.x.toFloat(), -camPos.y.toFloat(), -camPos.z.toFloat())
        scratchMVP.clear()
        worldProjMatrix.get(scratchMVP)
        mv.get(scratchMVP)
        scratchIdentity.get(scratchMVP)
        scratchMVP.flip()

        scratchParams.clear()
        scratchParams.putFloat(colorR).putFloat(colorG).putFloat(colorB).putFloat(colorA)
        scratchParams.putFloat(glowIntensity).putFloat(rimPower).putFloat(fresnelPower).putFloat(pulseSpeed)
        scratchParams.flip()

        val numQuads = vertexCount / 4
        val indexCount = numQuads * 6
        if (scratchIndexData.capacity() < indexCount * 2)
            scratchIndexData = ByteBuffer.allocateDirect(indexCount * 2).order(ByteOrder.nativeOrder())
        scratchIndexData.clear()
        for (i in 0 until numQuads) {
            val base = i * 4
            scratchIndexData.putShort(base.toShort()).putShort((base + 1).toShort()).putShort((base + 2).toShort())
            scratchIndexData.putShort(base.toShort()).putShort((base + 2).toShort()).putShort((base + 3).toShort())
        }
        scratchIndexData.flip()

        val device = RenderSystem.getDevice()
        val encoder = device.createCommandEncoder()
        encoder.writeToBuffer(entityGlowVBO!!.slice(), scratchVertex)
        encoder.writeToBuffer(entityGlowMVP!!.slice(), scratchMVP)
        encoder.writeToBuffer(entityGlowParams!!.slice(), scratchParams)
        encoder.writeToBuffer(entityGlowIBO!!.slice(), scratchIndexData)

        val pass = encoder.createRenderPass({ -> "aporia:entity_glow" }, colorView, OptionalInt.empty(), depthView, OptionalDouble.empty())
        pass.use {
            it.setPipeline(pipeline)
            it.setUniform("ModelViewProj", entityGlowMVP!!.slice())
            it.setUniform("DrawParams", entityGlowParams!!.slice())
            it.setVertexBuffer(0, entityGlowVBO!!)
            it.setIndexBuffer(entityGlowIBO!!, VertexFormat.IndexType.SHORT)
            it.drawIndexed(0, 0, indexCount, 0)
        }
    }

    fun drawTestQuad(
        colorR: Float, colorG: Float, colorB: Float, colorA: Float,
        glowIntensity: Float, rimPower: Float, fresnelPower: Float, pulseSpeed: Float
    ) {
        val pipeline = entityGlowPipeline ?: run { Logger.warn("drawTestQuad: pipeline null"); return }
        val mc2 = Minecraft.getInstance()
        val colorView = mc2.mainRenderTarget.colorTextureView ?: return
        val depthView = mc2.mainRenderTarget.depthTextureView

        val stride = DefaultVertexFormat.NEW_ENTITY.vertexSize
        val vCount = 4; val iCount = 6

        scratchVertex.clear()
        if (scratchVertex.capacity() < vCount * stride)
            scratchVertex = ByteBuffer.allocateDirect(vCount * stride).order(ByteOrder.nativeOrder())
        for (i in 0 until vCount) {
            val x = if (i == 0 || i == 3) -1.0f else 1.0f
            val y = if (i == 0 || i == 1) -1.0f else 1.0f
            scratchVertex.putFloat(x).putFloat(y).putFloat(-5.0f)
            scratchVertex.putInt(-1)
            scratchVertex.putFloat(i * 0.333f).putFloat(i * 0.333f)
            scratchVertex.putShort(15728880.toShort()).putShort(15728880.toShort())
            scratchVertex.putShort(0).putShort(0)
            scratchVertex.put(0).put(0).put(127).put(0)
        }
        scratchVertex.flip()

        val fov = Math.toRadians(70.0).toFloat()
        val aspect = Minecraft.getInstance().window.width.toFloat() / Minecraft.getInstance().window.height.toFloat()
        scratchCombined.setPerspective(fov, aspect, 0.1f, 1000.0f)
        scratchMVP.clear()
        scratchCombined.get(scratchMVP)
        scratchIdentity.get(scratchMVP)
        scratchIdentity.get(scratchMVP)
        scratchMVP.flip()

        scratchParams.clear()
        scratchParams.putFloat(colorR).putFloat(colorG).putFloat(colorB).putFloat(colorA)
        scratchParams.putFloat(glowIntensity).putFloat(rimPower).putFloat(fresnelPower).putFloat(pulseSpeed)
        scratchParams.flip()

        if (scratchIndexData.capacity() < iCount * 2)
            scratchIndexData = ByteBuffer.allocateDirect(iCount * 2).order(ByteOrder.nativeOrder())
        scratchIndexData.clear()
        scratchIndexData.putShort(0).putShort(1).putShort(2)
        scratchIndexData.putShort(0).putShort(2).putShort(3)
        scratchIndexData.flip()

        val device = RenderSystem.getDevice()
        val encoder = device.createCommandEncoder()
        encoder.writeToBuffer(entityGlowVBO!!.slice(), scratchVertex)
        encoder.writeToBuffer(entityGlowMVP!!.slice(), scratchMVP)
        encoder.writeToBuffer(entityGlowParams!!.slice(), scratchParams)
        encoder.writeToBuffer(entityGlowIBO!!.slice(), scratchIndexData)

        val pass = encoder.createRenderPass({ -> "aporia:test_quad" }, colorView, OptionalInt.empty(), depthView, OptionalDouble.empty())
        pass.use {
            it.setPipeline(pipeline)
            it.setUniform("ModelViewProj", entityGlowMVP!!.slice())
            it.setUniform("DrawParams", entityGlowParams!!.slice())
            it.setVertexBuffer(0, entityGlowVBO!!)
            it.setIndexBuffer(entityGlowIBO!!, VertexFormat.IndexType.SHORT)
            it.drawIndexed(0, 0, iCount, 0)
        }
    }
}
